rust/src/io/external_manifest/commit.rs

```diff
@@ -0,0 +1,438 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Trait for external manifest handler.
+//!
+//! This trait abstracts an external storage with put_if_not_exists semantics.
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use object_store::path::Path;
+use snafu::{location, Location};
+
+use crate::dataset::{latest_manifest_path, manifest_path};
+use crate::format::{Index, Manifest};
+use crate::io::commit::{CommitError, CommitHandler, ManifestWriter};
+use crate::io::{read_manifest, ObjectStore};
+
+use crate::{Error, Result};
+
+/// External manifest store
+///
+/// This trait abstracts an external storage for source of truth for manifests.
+/// The storge is expected to remember (uri, version) -> manifest_path
+/// and able to run transactions on the manifest_path.
+/// **only supports absolute paths**
+#[async_trait]
+trait ExternalManifestStore: std::fmt::Debug + Send + Sync {
```

Comment:
Are we going to consolidate "External vs Local" ManifestStore?

Should the trait to just be `ManifestStore`

Reply:
Actually what is the different between `ManifestStore` to a traditional Catalog

Reply:
added comment to explain the difference. TL;DR: this is called "external" is because it is expected to always sync back to object store.